installShellFiles: Enhance installShellCompletion

lilyball opened the pull request:
###### Motivation for this change
Some packages don't bundle shell completions as files in their source tree but instead synthesize them by invoking the built tool with specific flags. This enhances the `installShellCompletion` command such that it can install these completions without having to pipe them to a file first. To do this, it now supports a named pipe as a path, such as produced by `<(cmd)`. It also now has a convenience flag to synthesize the name for all files based on the shell.

Examples:

```sh
installShellCompletion --cmd foobar \
  --bash <($out/bin/foobar --bash-completion) \
  --fish <($out/bin/foobar --fish-completion) \
  --zsh <($out/bin/foobar --zsh-completion)
```

Fixes #83284

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [X] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

I tested one package that uses `installShellCompletion` (`ffsend`) and it produced the same output. I also manually tested all of the new functionality, including error cases.

Commit:
installShellFiles: Enhance installShellCompletion

Teach installShellCompletion how to install completions from a named
pipe. Also add a convenience flag `--cmd NAME` that synthesizes the name
for each completion instead of requiring repeated `--name` flags.

Usage looks something like

    installShellCompletion --cmd foobar \
      --bash <($out/bin/foobar --bash-completion) \
      --fish <($out/bin/foobar --fish-completion) \
      --zsh <($out/bin/foobar --zsh-completion)

Fixes #83284

lilyball commented:
I'd love to write a test suite for `installShellCompletion` but I'm not aware of any existing means to do such a thing.

lilyball commented:
#82964 is an example of a PR that could make use of this functionality. That PR has

```bash
    $PSC_PACKAGE --bash-completion-script $PSC_PACKAGE > psc-package.bash
     $PSC_PACKAGE --fish-completion-script $PSC_PACKAGE > psc-package.fish
     $PSC_PACKAGE --zsh-completion-script $PSC_PACKAGE > _psc-package
     installShellCompletion \
       psc-package.{bash,fish} \
       --zsh _psc-package
```

With this enhancement it could be rewritten as

```bash
    installShellCompletion --cmd psc-package \
      --bash --exec "$PSC_PACKAGE --bash-completion-script $PSC_PACKAGE" \
      --fish --exec "$PSC_PACKAGE --fish-completion-script $PSC_PACKAGE" \
      --zsh --exec "$PSC_PACKAGE --zsh-completion-script $PSC_PACKAGE"
```

rycee commented:
I haven't tried it but wouldn't something like
```sh
installShellCompletion --bash --name foobar.bash <($out/bin/foobar --bash-completion)
```
work?

lilyball commented:
I forgot that was a feature of bash. However, trying it right now on macOS gets me

```
install: skipping file '/dev/fd/63', as it was replaced while being copied
```

So at the very least `installShellCompletion` would have to be changed to use something other than `install`. Curiously, `install -T <(echo hi) x` works just fine on NixOS, it's just macOS where it's failing.

rycee commented:
Ah, didn't know that macos would have problems with that. When I wrote the comment I tried using `install` on NixOS using the `<(…)` and since that worked I figured the `installShellCompletion` would work.

Perhaps `<(…)` would work also on macOS if the `install` command was replaced by
```sh
mkdir -p "$outDir" \
  && cat "$arg" > "$outPath"
  && chmod 644 "$outPath"
```
?

In any case, I'm not against the `-` implementation but figured this solution might be worth considering since it's quite simple. If it can't easily be made to work on macOS then it's a non-starter, though.

lilyball commented:
I've updated this PR to remove the `--exec` flag in favor of `<(cmd)`. It now detects named pipes and used `cat` to read them, still using `install` for regular paths. I didn't bother adding the `chmod 644` since `-rw-r--r--` seems to be the default permissions anyway, it's just there on the `install` because it's easy to be specific there.

Commit:
psc-package: Adopt new installShellCompletion syntax

lilyball commented:
I just added a second commit that updates `psc-package` to use this new syntax.

Profpatsch commented:
I’d love to have this new functionality, but I’m really uncomfortable with the amount of complexity it adds to the bash code (and all that command line parsing logic and interacting command line arguments, halp).

I’d feel a lot better if there was a test suite and if possible a list of packages that use this feature and whose output could be checked to contain the completions in the right location.

Maybe we can reduce the scope and just add the simplest subset of this PR first, instead of doing it all in one go?

lilyball commented:
@Profpatsch I'd really love a test suite too, but I couldn't find any precedent for writing test suites on package setup hooks so I'm really not sure how to go about doing it. I did manually verify that `ffsend` still installs its completions correctly (which exercises the pre-existing functionality) and the second commit updates `psc-package` to use most of the new functionality.

> Maybe we can reduce the scope and just add the simplest subset of this PR first, instead of doing it all in one go?

I could remove stdin support, but that doesn't really save much. A more compelling argument for removing it is "why have 2 ways to do the same thing?", but I'm conflicted here because I feel like using stdin is perhaps more obvious to people than using `<(cmd)`. Still, perhaps the conceptual simplification is worth it.

FWIW I wrote this package setup hook in the first place (see #65211).

lilyball commented:
I went ahead and [stripped out the stdin support](https://github.com/NixOS/nixpkgs/compare/2c18eabc6314db928ece4bb40e999fc65e57b275..40e20f74828ada96d6ed9484ca6b81ea92ce60e8). Not a huge difference to the patch, but I did it for the conceptual simplification.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review-may-2019/3032/139

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review-may-2019/3032/152

Ericson2314 commented:
@lilyball this is looking good to me, though I am no too familiar with this stuff. There are tests in `pkgs/tests` for other `pkgs/build-support` items, so it could go there.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review-may-2019/3032/153

lilyball commented:
`pkgs/test/patch-shebangs/default.nix` looks like a good template. I'll try to find the time soon to write some tests like that.

lilyball committed:
Fix doc comment for installManPage

lilyball committed:
installShellFiles: Add test suite

lilyball commented:
I just added a suite of tests at `tests.install-shell-files`.

lilyball commented:
I didn't add the tests anywhere to be invoked automatically though.

Luflosi commented:
According to https://nixos.org/nixpkgs/manual/#submitting-changes-staging-branch, PRs with mass-rebuilds should be made to the staging branch. Does this apply to this PR?

lilyball commented:
I'm surprised at how much stuff indirectly depends on this. I guess cargo is using it for something, not sure what other non-leaf packages use it. I'm also wondering why ofborg slapped the more extreme labels on here when I added the test; AFAIK the addition of the test file shouldn't affect anything at all unless you explicitly build the test by name (should I be adding it to `installShellFiles.passthru.tests`? I'm not familiar with that).

Profpatsch commented:
Looks like these tests are called in the release tarball: https://github.com/profpatsch/nixpkgs/blob/current-build/pkgs/top-level/release.nix#L87-L125

I don’t think your test it’s currently called anywhere, so the test should not increase the number of reverse dependencies.

Profpatsch commented:
I’m a bit surprised by the number of rebuilds ofborg shows, grepping for `installShellCompletion` I just see a few dozen uses in all of nixpkgs.
Plus the way stdenv is structured, changing a hook should only change the packages that pull that hook in; maybe a language packageset pulls it in for all of its packages and I missed that.

lilyball commented:
It looks as though `darwin.cctools` (which is part of the darwin stdenv) has sprouted a dependency on `installShellFiles` at some point between this PR's creation and current master. This is why the PR suddenly requires a rebuild of the entire darwin world.

Profpatsch commented:
You’ll still need to include your test in the release tests though, for it to be picked up.

An interesting phenomenon is that ofborg picks up the `passthru.tests` of reverse dependencies.

lilyball commented:
I wasn't sure if I should put it in the release file or not. Are you recommending that I do? Or should I just put it in `installShellFiles.passthru.tests`?

lilyball commented:
Given that `darwin.cctools` now depends on this, which means all of stdenv does, do I need to point this at staging? FWIW `darwin.cctools` just uses `installManPage` which this PR doesn't touch.

Profpatsch commented:
I’d go for `staging`

On Thu, May 21, 2020 at 1:20 AM Lily Ballard <notifications@github.com>
wrote:

> Given that darwin.cctools now depends on this, which means all of stdenv
> does, do I need to point this at staging? FWIW darwin.cctools just uses
> installManPage which this PR doesn't touch.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/83630#issuecomment-631788059>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAYB5ZSCHJNC32Q6HJJLBXLRSRQTHANCNFSM4LVXXWFQ>
> .
>

Profpatsch commented:
> I wasn't sure if I should put it in the release file or not. Are you recommending that I do?

I guess it makes sense? I don’t have a good overview on when these are run, probably before advancing channels.



> Or should I just put it in `installShellFiles.passthru.tests`?

Ah, I wasn’t aware `installShellFiles` (or more generally anything created by `makeSetupHook` was an actual derivation.

In that case, let’s go for the `passthru.tests

Luflosi commented:
@lilyball Could you please resolve the conflict?
Do you have an idea, when this might be merged?

lilyball commented:
@Luflosi Sorry, I haven't had time to work on this in ages. I wanted to put the tests into `passthru.tests` so they'd actually be run. I'll try to find the time to come back to this soon.

blaggacao commented:
I would have had an interest in using this better sooner than later. Most Go tools behave according to what this PR tries to solve.

lilyball commented:
I can't promise it but I will do my best to revisit this today.

lilyball commented:
I rebased onto `staging`, fixed the merge conflict, and added `passthru.tests`. For the tests I had to use `overrideAttrs` as `makeSetupHook` has no direct support for this; I can't test locally because of the mass rebuild but I'm assuming it will work.

Profpatsch commented:
It has tests, it adds some features, doesn’t seem to break the existing stuff, LGTM!

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/how-many-people-are-paid-to-work-on-nix-nixpkgs/8307/79
